File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// class Solution {
//     TreeNode* fun(TreeNode* root, TreeNode* p, TreeNode* q) {
//         if(root == nullptr || root == p || root == q) return root;

//         TreeNode* l = fun(root->left, p, q);
//         TreeNode* r = fun(root->right, p, q);

//         if(l != nullptr && r != nullptr) return root; // this means this curr node is LCS

//         return l ? l : r;
//     }
// public:
//     TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
//         return fun(root, p, q);
//     }
// };


class Solution {
    TreeNode* fun(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == nullptr) return root;

        if(p->val < root->val && q->val < root->val) {
            return fun(root->left, p, q);
        } else if(p->val > root->val && q->val > root->val) {
            return fun(root->right, p, q);
        } else {
            return root;
        }

        return root;
    }
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return fun(root, p, q);
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int ans = 0;
    private int fun(TreeNode root) {
        if(root == null) return 0;

        int l = 0, r = 0;
        if(root.left != null) l = 1 + fun(root.left);
        if(root.right != null) r = 1 + fun(root.right);

        ans = Math.max(ans, l + r);

        return Math.max(l, r);
    }
    public int diameterOfBinaryTree(TreeNode root) {
        fun(root);
        return ans;
    }
}
